Changed logging libs from GLOG to ABSEIL

ctwarog-ross opened the pull request:
Closes Issue #1016

Now uses ABSL with custom FileSink that specifies file location UT always log to ABSL_LOGGING_DIR specified in cmake Examples use relative ./logs folder as default unless specified with new --logdir arg --silent still silences all logs except errors
Release mode only prints to logs
Debug prints to std and logs
Added unit test for log coverage

ctwarog-ross committed:
Changed logging libs from GLOG to ABSEIL

Closes Issue #1016

Now uses ABSL with custom FileSink that specifies file location
UT always log to ABSL_LOGGING_DIR specified in cmake
Examples use relative ./logs folder as default unless specified with new --logdir arg
--silent still silences all logs except errors
Release mode only prints to logs
Debug prints to std and logs
Added unit test for log coverage

ctwarog-ross commented on sdks/cpp/common/include/Logger.h:
```diff
@@ -38,71 +38,80 @@
+
   std::ostringstream stream_;
+  std::unique_ptr<FileLogSink> fileLogSink_ = nullptr;
 };
```
newline, bad christian

ctwarog-ross commented on sdks/cpp/connections/REST/examples/asset_request_REST/asset_request_REST.cpp:
```diff
@@ -191,15 +191,10 @@ void RunRESTServer() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
again, bad

ctwarog-ross commented on sdks/cpp/connections/REST/examples/discovery_REST/discovery_REST.cpp:
```diff
@@ -108,18 +108,13 @@ void RunRESTServer() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
buddy

ctwarog-ross commented on sdks/cpp/connections/REST/examples/structs_with_authz_JSON_REST/structs_with_authz_JSON_REST.cpp:
```diff
@@ -145,15 +145,10 @@ void RunRESTServer() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 } 
```
this one wasnt me

ctwarog-ross commented on sdks/cpp/connections/REST/examples/structs_with_authz_REST/structs_with_authz_REST.cpp:
```diff
@@ -145,15 +145,10 @@ void RunRESTServer() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 } 
```
^

abrown-RV commented on sdks/cpp/common/include/Logger.h:
```diff
@@ -38,71 +38,80 @@
+  * @param appName The name of the application.
+  * @param logDir The directory where log files will be created.
+  */
+  static void init(const std::string& appName, const std::string& logDir = ABSL_LOG_DIR);
```
suggest using `LOGGING_DIR` so if we change logging libraries in the future, we can reuse without having to go through all the cmake files again.

abrown-RV commented on sdks/cpp/common/src/Logger.cpp:
```diff
@@ -0,0 +1,86 @@
+    });
+}
+
+void Logger::init(int argc, char** argv, const std::string& appName) {
```
want to contemplate using the `ABSL_FLAG` system? since this is all abseil now? Benefit being I think that you'd no longer need to take `argc` and `argv` so you'd only need the one `init`. `ABSL_FLAG`s seem to be able to magically just read the args.

Maybe even add to `SharedFlags` to continue with the pattern?

ctwarog-ross replied:
we need all the args because of the added functionality my wrapper provides. For ex, we parse a --silent arg

abrown-RV commented on sdks/cpp/common/src/Logger.cpp:
```diff
@@ -0,0 +1,86 @@
+            std::filesystem::create_directories(logDir);
+        }
+
+#ifdef NDEBUG
```
I think we should give sdk users the option of logging to console in prod if they want. But maybe something to look at when we actually talk about making this something others might use. So feel free to ignore, we're already talked about logging to strout/err a bit to death XD

abrown-RV commented on sdks/cpp/connections/REST/examples/use_commands_REST/use_commands_REST.cpp:
```diff
@@ -184,18 +184,13 @@ void defineCommands() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 } 
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/REST/examples/use_menus_REST/use_menus_REST.cpp:
```diff
@@ -152,15 +152,10 @@ void RunRESTServer() {
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 } 
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/one_of_everything/one_of_everything.cpp:
```diff
@@ -362,7 +362,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/status_update/status_update.cpp:
```diff
@@ -241,7 +241,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/structs_with_authz/structs_with_authz.cpp:
```diff
@@ -179,7 +179,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/structs_with_authz_yaml/structs_with_authz_yaml.cpp:
```diff
@@ -176,7 +176,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/use_commands/use_commands.cpp:
```diff
@@ -279,7 +279,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/use_menus/use_menus.cpp:
```diff
@@ -178,7 +178,5 @@ int main(int argc, char* argv[])
-    // Shutdown Google Logging
-    google::ShutdownGoogleLogging();
     return 0;
 }
```
```suggestion
}

```

abrown-RV commented on sdks/cpp/connections/REST/src/controllers/DeviceRequest.cpp:
```diff
@@ -87,5 +87,5 @@ void DeviceRequest::proceed() {
     writeConsole_(CallStatus::kFinish, socket_.is_open());
-    DEBUG_LOG << "DeviceRequest[" << objectId_ << "] finished\n";
+    LOG(INFO) << "DeviceRequest[" << objectId_ << "] finished\n";
 }
```
```suggestion
}

```

abrown-RV commented on unittests/cpp/common/tests/ConnectionQueue_test.cpp:
```diff
@@ -51,11 +51,10 @@ class ConnectionQueueTest : public testing::Test {
     // Set up and tear down Google Logging
     static void SetUpTestSuite() {
-        Logger::StartLogging("ConnectionQueueTest");
+      Logger::init("ConnectionQueueTest", ABSL_LOG_DIR);
```
Why does this on get `ABSL_LOG_DIR`?

Indentation

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,72 @@
+            break;
+        }
+    }
+}
```
```suggestion
}

```

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,72 @@
+    std::string logDir = ABSL_LOG_DIR;
+    std::string logFilePattern = "LoggerTest_"; // Log file name starts with this
+    bool foundLogFile = false;
+    for (const auto& entry : std::filesystem::directory_iterator(logDir)) {
```
This test verification design makes me a little uncomfortable. As written, if for whatever reason there was a failing test in the logging dir. You'd have to go an manually delete it before tests would pass in the future. I think it would be better to clear out all `LoggerTest_` log files before calling `Logger::init`, or you could make this a fixtured test and then put it in the teardown (so it runs even if the test fails).

Another though I have is to not use `ABSL_LOG_DIR`, but instead a separate logging dir for these test(s). Since its a param in the init. Would make it easier to clean up logs from past test runs. What I like to do is make the directory and put a .gitignore file with a `*` in it in the dir. This works nicely because the .gitignore file will ignore all other files in the dir, and also gets around git's limitation of not being able to commit empty dirs by giving it a file in the dir to commit!

I do appreciate the desire to test the maybe more common and full path of if `init` uses the `ABSL_LOG_DIR` as the default. So since there's only one test possible, maybe ignore my last paragraph.

I do still think cleaning up test artifacts between tests if they will effect the pass/fail of the test to be important.

ctwarog-ross replied:
No the last paragraph is valid, making a folder in the ut dir for this is a smart idea, as well as cleaning up artifacts in the teardown.

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,72 @@
+    for (const auto& entry : std::filesystem::directory_iterator(logDir)) {
+        if (entry.is_regular_file() && entry.path().filename().string().find(logFilePattern) == 0) {
+            foundLogFile = true;
+            std::ifstream logFile(entry.path());
```
Could also do `EXPECT_THAT(entry.path().filename().string(), MatchesRegex(pattern));` to assert the filename has the expected pattern with a formatted time, etc.

abrown-RV reviewed (changes requested):
Mostly newlines at the end of files if we're going to be picky about those.

Some actual thoughts as well on the real code.

ctwarog-ross committed:
Applied pr feedback

cmd args now use ABSL flag system
ABSL now used from specified version built from source as GRPC submodule
logging test validates file name regex

abrown-RV commented on .devcontainer/devcontainer.dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM ghcr.io/rossvideo/catena-toolchain-cpp:latest
+FROM testing:latest
```
```suggestion
FROM ghcr.io/rossvideo/catena-toolchain-cpp:latest
```

ctwarog-ross replied:
doh!

abrown-RV commented on .devcontainer/toolchain-cpp.dockerfile:
```diff
@@ -85,11 +85,9 @@ RUN . /root/toolchain.env \
 # Install Python and gcovr for coverage reports
 RUN apt-get update && apt-get install --no-install-recommends -y python3-pip \
-    && pip3 install --no-cache-dir -r /root/requirements.txt --break-system-packages \
+    && pip3 install --no-cache-dir gcovr --break-system-packages \
```
```suggestion
    && pip3 install --no-cache-dir -r /root/requirements.txt--break-system-packages \
```

abrown-RV commented on .devcontainer/toolchain-cpp.env:
```diff
@@ -25,3 +25,4 @@ LIBCURL_VERSION=8.5.0-2ubuntu10.6
 AVAHI_CLIENT_VERSION=0.8-13ubuntu6
 AVAHI_COMMON_VERSION=0.8-13ubuntu6
 PKG_CONFIG_VERSION=1.8.1-2build1
+ABSEIL_VERSION=4a2c633
```
Points to [20240116.rc2](https://github.com/abseil/abseil-cpp/releases/tag/20240116.rc2). Probably want to not be on a Release Candidate, but a full release?

Suggest putting the version tag in a comment as well since commit hashes are so opaque. 
Since you're running git checkout in the submodule it doesn't have to be a commit hash, could be the tag.

ctwarog-ross replied:
I tried the latest release, but it wouldn't build due to incompatibility with grpc. I ended up choosing this version as thats the hash that the version of grpc we use has checked out

ctwarog-ross replied:
so essentially we're hardcoded locked into this combination of grpc and abseil

abrown-RV replied:
gotcha, can you lock it to the version tag instead of the hash for readability?

abrown-RV commented on sdks/cpp/cmake/ProjectConfig.cmake:
```diff
@@ -21,24 +21,15 @@ function(setup_project_directories)
 # Setup logging configuration  
-function(setup_logging_config CATENA_ROOT_DIR)
+function(setup_logging_config)
     # Set default log directory - respect command line override
```
Do we want to have a default `LOG_DIR`? What happens if it's not set, does `add_compile_definitions` still work? I guess it would inject `LOG_DIR=""`. Which the code probably is fine with?

ctwarog-ross replied:
the log dir is passed in cmd args in the dockerfile. My thought was if someone wants to go build it their own way, they can change it there. I will add that iif they don't, the default location is set to Catena/logs in the cmake files anyway.

abrown-RV replied:
I'm only familiar with using the `option` cmake directive, is there another way?

abrown-RV commented on sdks/cpp/common/src/Logger.cpp:
```diff
@@ -0,0 +1,76 @@
+    std::call_once(flag, [&]() {
+        absl::InitializeLog();
+
+        auto test = absl::GetFlag(FLAGS_log_dir);
```
unused? Guessing it was a gdb watch?

abrown-RV commented on sdks/cpp/common/src/Logger.cpp:
```diff
@@ -0,0 +1,76 @@
+        absl::InitializeLog();
+
+        auto test = absl::GetFlag(FLAGS_log_dir);
+        // Optionally set log dir via env or here
```
```suggestion
        // Create the log dir if it doesn't exist
```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/audiodeck/audiodeck.cpp:
```diff
@@ -168,18 +168,14 @@ void RunRPCServer(std::string addr)
     std::string addr;
-    absl::SetProgramUsageMessage("Runs the Catena Service");
     absl::ParseCommandLine(argc, argv);
+    Logger::init("structs_with_authz_REST");
```
```suggestion
    Logger::init("audiodeck");
```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/audiodeck_yaml/audiodeck_yaml.cpp:
```diff
@@ -165,18 +165,14 @@ void RunRPCServer(std::string addr)
-    absl::SetProgramUsageMessage("Runs the Catena Service");
     absl::ParseCommandLine(argc, argv);
-  
+    Logger::init("structs_with_authz_REST");
```
```suggestion
    Logger::init("audiodeck_yaml");
```

abrown-RV commented on sdks/cpp/connections/gRPC/examples/audiodeck/audiodeck.cpp:
```diff
@@ -168,18 +168,14 @@ void RunRPCServer(std::string addr)
-    Logger::StartLogging(argc, argv);
-
     std::string addr;
-    absl::SetProgramUsageMessage("Runs the Catena Service");
```
Why remove?

ctwarog-ross replied:
doesnt add anything to the log

abrown-RV replied:
Its not for logging. Its part of the absl flags system. It prints that message when the user runs the example with `--help`

abrown-RV commented on sdks/cpp/connections/REST/examples/audiodeck_JSON_REST/audiodeck_JSON_REST.cpp:
```diff
@@ -145,15 +145,11 @@ void RunRESTServer() {
-    absl::SetProgramUsageMessage("Runs the Catena Service");
     absl::ParseCommandLine(argc, argv);
-    
+    Logger::init("structs_with_authz_JSON_REST");
```
```suggestion
    Logger::init("audiodeck_JSON_REST");
```

abrown-RV commented on sdks/cpp/connections/REST/examples/audiodeck_REST/audiodeck_REST.cpp:
```diff
@@ -145,15 +145,11 @@ void RunRESTServer() {
-    absl::SetProgramUsageMessage("Runs the Catena Service");
     absl::ParseCommandLine(argc, argv);
-    
+    Logger::init("structs_with_authz_REST");
```
```suggestion
    Logger::init("audiodeck_REST");
```

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,84 @@
+class LoggerTest : public ::testing::Test {
+    void SetUp() override {
+        absl::SetFlag(&FLAGS_log_dir, UNITTEST_LOG_DIR + std::string("/logger"));
+        Logger::init("LoggerTest");
```
I realized I misspoke last round, I actually intentionally like to do the clean in the setup, that way, if there's a test that doesn't pass, the artifacts are left behind to poke around at them.

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,84 @@
+    bool foundLogFile = false;
+    for (const auto& entry : std::filesystem::directory_iterator(UNITTEST_LOG_DIR + std::string("/logger"))) {
+        //get the file that ends with _LoggerTest
+        if (entry.is_regular_file() && entry.path().filename().string().find(logFilePattern) == entry.path().filename().string().length() - logFilePattern.length()) {
```
Now that you're in an empty subdir for this test, don't just skip non matching files. Any extra files are a failed test, cause it should only make the one. I think you still need the loop, since you don't concretely know the name of the log file cause of the time, but the loop can just expect the regex pattern, and then also expect the count to be one. Unless there's a fancy iter way to just get the first element out of the iter without a loop. Then just do that and then expect the iter is at the end.

ctwarog-ross replied:
I could also just check if it has more than 1 file?

abrown-RV replied:
totally

abrown-RV commented on unittests/cpp/gRPC/GRPCTest.h:
```diff
@@ -58,6 +58,7 @@
 // common
 #include <Status.h>
 #include <Logger.h>
+#include <SharedFlags.h>
```
```suggestion
```

ctwarog-ross replied:
it needs this to be able to set the unit test dir flag

abrown-RV replied:
Where? this file doesn't touch the flag. All the files that include GRPCTest.h `#include <SharedFlags.h>` themselves.

ctwarog-ross replied:
gotchu, I must haver forgotten to remove it then.

ctwarog-ross replied:
nope, just looked through and none of the files that include grpctest.h include sharedflags. Leaving as is

abrown-RV commented on unittests/cpp/common/tests/Logger_test.cpp:
```diff
@@ -0,0 +1,84 @@
+            std::ifstream logFile(entry.path());
+            std::string logContent((std::istreambuf_iterator<char>(logFile)), std::istreambuf_iterator<char>());
+            EXPECT_NE(logContent.find("This is an info log message."), std::string::npos);
+            EXPECT_NE(logContent.find("This is a warning log message."), std::string::npos);
```
Check for error log message.

abrown-RV reviewed (commented):
Super cool `absl::SetFlag` exists. Great for testing!

ctwarog-ross committed:
applied PR feedback

ctwarog-ross committed:
added back SetPRogramUsageMessage

ctwarog-ross commented:
BLOCKED by https://github.com/rossvideo/Catena/pull/1090

abrown-RV committed:
revert the device model changes
